src/models/kv_cache.cpp

```diff
@@ -213,25 +267,83 @@ void KV_Cache::Add() {
   }
 }
 
-void KV_Cache::Update(DeviceSpan<int32_t> beam_indices, int current_length) {
+void KV_Cache::Update(DeviceSpan<int32_t> beam_indices, int total_length) {
   // If we're sharing past & present buffers there is nothing to do here, so early exit
   if (past_present_share_buffer_)
     return;
 
-  for (int i = 0; i < layer_count_ * 2; i++) {
-    if (beam_indices.empty()) {
-      pasts_[i] = std::move(presents_[i]);
-    } else {
-      PickPastState(beam_indices, i);
+  if (!is_first_update_) {
+    for (int i = 0; i < layer_count_ * 2; i++) {
+      if (beam_indices.empty()) {
+        pasts_[i] = std::move(presents_[i]);
+      } else {
+        PickPastState(beam_indices, i);
+      }
+      state_.inputs_[input_index_ + i] = pasts_[i].get();
     }
-    state_.inputs_[input_index_ + i] = pasts_[i].get();
   }
 
-  shape_[2] = current_length;
+  shape_[2] = total_length;
   for (int i = 0; i < layer_count_ * 2; i++) {
     presents_[i] = OrtValue::CreateTensor(*model_.allocator_kvcache_, shape_, type_);
     state_.outputs_[output_index_ + i] = presents_[i].get();
   }
+
+  is_first_update_ = false;
+}
+
+void KV_Cache::RewindTo(size_t index) {
+  if (past_present_share_buffer_) {
+    return;
+  } else if (shape_[2] <= static_cast<int>(index)) {
+    throw std::runtime_error("Requested length of rewind is greater than the current length.");
+  }
+
+  is_first_update_ = true;
+  if (index == 0) {
+    for (int i = 0; i < layer_count_ * 2; i++) {
+      pasts_[i] = nullptr;
+      state_.inputs_[input_index_ + i] = empty_past_.get();
+    }
+  } else if (type_ == Ort::TypeToTensorType<float>) {
+    RewindPastTensorsTo<float>(index);
+  } else {
+    RewindPastTensorsTo<Ort::Float16_t>(index);
+  }
+}
+
+template <typename T>
+void KV_Cache::RewindPastTensorsTo(size_t index) {
+  assert(index > 0 && shape_[2] >= index && !past_present_share_buffer_);
+  std::array<int64_t, 4> new_shape = shape_;
+  new_shape[2] = static_cast<int>(index);
+  auto batch_x_num_heads = new_shape[0] * new_shape[1];
+  auto new_length_x_head_size = new_shape[2] * new_shape[3];
+  auto old_length_x_head_size = shape_[2] * new_shape[3];
+  shape_[2] = new_shape[2];
+
+  for (int i = 0; i < layer_count_ * 2; i++) {
+    OrtValue& present = *presents_[i];
+    std::unique_ptr<OrtValue> past = OrtValue::CreateTensor(*model_.allocator_kvcache_, shape_, type_);
+    for (int j = 0; j < batch_x_num_heads; j++) {
+      auto present_data = present.GetTensorData<T>() + j * old_length_x_head_size;
+      auto past_data = past->GetTensorMutableData<T>() + j * new_length_x_head_size;
+#if USE_CUDA
+      if (model_.device_type_ == DeviceType::CUDA) {
+        cudaMemcpyAsync(past_data, present_data, new_length_x_head_size * sizeof(T), cudaMemcpyDeviceToDevice, model_.cuda_stream_);
+      } else
+#elif USE_DML
```

Comment:
It is possible that we may enable use cuda and use dml together. Do you want to handle this differently here?